src/Chapter3.hs

```diff
@@ -1005,8 +1111,85 @@ properties using typeclasses, but they are different data types in the end.
 Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
-
-
+data Knight'Action = KAttack
+                   | HealthPotion
+                   | IncreaseArmor deriving (Show)
+data Monster'Action = MAttack
+                    | RunAway deriving (Show)
+
+-- data Action = KAction Knight'Action | MAction Monster'Action deriving (Show)
+data Knight' = Knight' { knight'Health  :: Health
+                       , knight'Attack  :: Attack
+                       , knight'Armor   :: Armor
+                       , knight'Actions :: [Knight'Action]
+                       } deriving (Show)
+data Monster' = Monster' { monster'Health  :: Health
+                         , monster'Attack  :: Attack
+                         , monster'Actions :: [Monster'Action]
+                         } deriving (Show)
+
+class (Show a) => Fighter a where
+  performAction :: (Fighter b) => a -> b -> (a, b)
+  receiveDamage :: a -> Attack -> a
+  getHealth :: a -> Int
+
+instance Fighter Knight' where
+  performAction :: (Fighter enemy) => Knight' -> enemy -> (Knight', enemy)
+  performAction knight@(Knight' { knight'Health = (Health kHealth)
+                                , knight'Attack = kAttack
+                                , knight'Armor  = (Armor  kArmor)
+                                , knight'Actions = (currentAction:actions)
+                                })
+                enemy =
+    case currentAction of
+      KAttack -> (knight', receiveDamage enemy kAttack)
+      HealthPotion -> (knight' { knight'Health = Health (kHealth + 20) }, enemy)
+      IncreaseArmor -> (knight' { knight'Armor = Armor (kArmor + 5) }, enemy)
+    where
+      knight' = knight { knight'Actions = actions ++ [currentAction] }
```

Comment:
This is quite slow to add one-by-one actions at the end of the list on each round. Remember trains from the previous Chapter?
It would be better to create Knights only with infinite lists of actions 🙂